opm/output/eclipse/Summary.cpp

```diff
@@ -0,0 +1,469 @@
+     * If other semantics when just gas is produced, this must be changed.
+     */
+    if( oil == 0 ) return 0;
+
+    return gas / oil;
+}
+
+inline double wgorh( const Well& w, size_t ts ) {
+    /* We do not support mixed injections, and gas/oil is undefined when oil is
+     * zero (i.e. pure gas injector), so always output 0 if this is an injector
+     */
+    if( w.isInjector( ts ) ) return 0;
+
+    const auto& p = w.getProductionProperties( ts );
+
+    return wgor( p.GasRate, p.OilRate );
+}
+
+enum class WT { wat, oil, gas };
+inline double prodrate( const Well& w, size_t timestep, WT wt, const double* dims ) {
+    if( !w.isProducer( timestep ) ) return 0;
+
+    const auto& p = w.getProductionProperties( timestep );
+    switch( wt ) {
+        case WT::wat: return convert( p.WaterRate, dim::liquid_surface_rate, dims );
+        case WT::oil: return convert( p.OilRate, dim::liquid_surface_rate, dims );
+        case WT::gas: return convert( p.GasRate, dim::gas_surface_rate, dims );
+    }
+
+    throw std::runtime_error( "Reached impossible state in prodrate" );
+}
+
+inline double prodvol( const Well& w, size_t timestep, WT wt, const double* dims ) {
+    const auto rate = prodrate( w, timestep, wt, dims );
+    return rate * convert( 1, dim::time, dims );
+}
+
+inline double injerate( const Well& w,
+                        size_t timestep,
+                        WellInjector::TypeEnum wt,
+                        const double* dims ) {
+
+    if( !w.isInjector( timestep ) ) return 0;
+    const auto& i = w.getInjectionProperties( timestep );
+
+    /* we don't support mixed injectors, so querying a water injector for
+     * gas rate gives 0.0
+     */
+    if( wt != i.injectorType ) return 0;
+
+    if( wt == WellInjector::GAS )
+        return convert( i.surfaceInjectionRate, dim::gas_surface_rate, dims );
+
+    return convert( i.surfaceInjectionRate, dim::liquid_surface_rate, dims );
+}
+
+inline double injevol( const Well& w, size_t timestep, WellInjector::TypeEnum wt, const double* dims ) {
+    const auto rate = injerate( w, timestep, wt, dims );
+    return rate * convert( 1, dim::time, dims );
+}
+
+inline double get_rate( const data::Well& w, rt phase, const double* dims ) {
+    const auto x = w.rates.get( phase, 0.0 );
+    switch( phase ) {
+        case rt::gas: return convert( x, dim::gas_surface_rate, dims );
+        default: return convert( x, dim::liquid_surface_rate, dims );
+    }
+}
+
+inline double get_vol( const data::Well& w, rt phase, const double* dims ) {
+    const auto x = w.rates.get( phase, 0.0 );
+    switch( phase ) {
+        case rt::gas: return convert( x, dim::gas_surface_volume, dims );
+        default: return convert( x, dim::liquid_surface_volume, dims );
+    }
+}
+
+inline double well_keywords( const smspec_node_type* node,
+                             const ecl_sum_tstep_type* prev,
+                             const double* dims,
```

Comment:
I know have given unclear signals regarding the `dims` pointer. The correct&trade; hierarchy is: "Unit system -> Conversion factors" - where the conversion factors are *indexed by dimension*. So we could (I am *not* suggesting this - just for demonstration) have:

```C++

class ConversionFactor
public:
  double getConversionFactor( const std::string& systemName , dim d)  {
          const auto& system = conversionFactors.at(system);
          double factor = system.at( d );
          return factor;
  }

private:
   std::map<std::string,std::map<dim , double> > conversionFactors;
}
```